File: src/core/main/FileSystem.cs
using System.Collections.Generic;
using System.IO;


namespace Mason
{
    public static class FileSystem
    {
        public delegate bool LocationPredicate(string location);

        public static IEnumerable<string> EnumerateDirectories(string location, LocationPredicate filter, LocationPredicate termninateCondition)
        {
            var root = Directory.GetDirectoryRoot(location);
            var current = location;
            do
            {
                var dir = current;
                current = Directory.GetParent(current).FullName;
                if (filter == null || filter(dir))
                {
                    yield return dir;
                }
                if (termninateCondition != null && termninateCondition(dir))
                {
                    break;
                }
            }
            while (current != root);
        }
        public static IEnumerable<string> EnumerateDirectories(string location, LocationPredicate filter)
        {
            return EnumerateDirectories(location, filter, null);
        }

        public static IEnumerable<string> SearchDirectories(string location, string pattern, SearchOption searchOption)
        {
            return Directory.GetFiles(location, pattern, searchOption);
        }
        public static IEnumerable<string> SearchDirectories(string location, string pattern) 
        { 
            return SearchDirectories(location, pattern, SearchOption.TopDirectoryOnly); 
        }
    }
}

File: src/foundation/main/FileSystem.fs
namespace Mason

open System
open System.IO

[<Obsolete>]
module FileSystem =
    type LocationPredicate = delegate of location: DirectoryInfo -> bool

    let always(s: DirectoryInfo) = true
    let never(s: DirectoryInfo) = false

    let rec recEnumerateDirectories(location: DirectoryInfo, filter: LocationPredicate, terminator: LocationPredicate) =
        match null2opt location with
        | Some loc ->
            if (terminator.Invoke(loc)) then []
            else if (filter.Invoke(loc)) then 
                if (loc = loc.Root) then [loc]
                else [loc] @ recEnumerateDirectories(loc.Parent, filter, terminator)
            else
                if (loc = location.Root) then [loc]
                else [loc] @ recEnumerateDirectories(loc.Parent, filter, terminator)
        | None -> []

    let EnumerateDirectories(location: string, filter: LocationPredicate, terminateCondition: LocationPredicate) = 
        let f = match null2opt filter with | Some fn -> fn | None -> LocationPredicate always
        let t = match null2opt terminateCondition with | Some fn -> fn | None -> LocationPredicate never
        match null2opt location with 
        | Some loc -> recEnumerateDirectories(new DirectoryInfo(loc), f, t) :> seq<DirectoryInfo>
        | None -> raise (ArgumentNullException("location"))

    let SearchDirectories(option: SearchOption, location: string, pattern: string) = 
        match null2opt location with 
        | Some loc ->
            match null2opt pattern with 
            | Some p -> DirectoryInfo(loc).GetFiles(p, option) :> seq<FileInfo>
            | None -> raise (ArgumentNullException("pattern"))
        | None -> raise (ArgumentNullException("location"))
        
            

